src/libos/crates/async-rt/src/time/instant.rs

```diff
@@ -0,0 +1,220 @@
+        })
+    }
+}
+
+impl Sub<Instant> for Instant {
+    type Output = Duration;
+
+    fn sub(self, other: Instant) -> Duration {
+        self.duration_since(other)
+    }
+}
+
+impl Add<Duration> for Instant {
+    type Output = Instant;
+
+    fn add(self, other: Duration) -> Instant {
+        self.checked_add(other)
+            .expect("overflow when adding duration to instant")
+    }
+}
+
+impl AddAssign<Duration> for Instant {
+    fn add_assign(&mut self, other: Duration) {
+        *self = *self + other;
+    }
+}
+
+impl Sub<Duration> for Instant {
+    type Output = Instant;
+
+    fn sub(self, other: Duration) -> Instant {
+        self.checked_sub(other)
+            .expect("overflow when subtracting duration from instant")
+    }
+}
+
+impl SubAssign<Duration> for Instant {
+    fn sub_assign(&mut self, other: Duration) {
+        *self = *self - other;
+    }
+}
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Ord, PartialOrd)]
+struct Timespec {
+    pub tv_sec: i64,
+    pub tv_nsec: i64,
+}
+
+impl Timespec {
+    pub fn zero() -> Self {
+        Self {
+            tv_sec: 0,
+            tv_nsec: 0,
+        }
+    }
+
+    pub fn sub_timespec(&self, other: &Timespec) -> Option<Duration> {
+        if self >= other {
+            if self.tv_nsec >= other.tv_nsec {
+                Some(Duration::new(
+                    (self.tv_sec - other.tv_sec) as u64,
+                    (self.tv_nsec - other.tv_nsec) as u32,
+                ))
+            } else {
+                Some(Duration::new(
+                    (self.tv_sec - 1 - other.tv_sec) as u64,
+                    self.tv_nsec as u32 + (NSEC_PER_SEC as u32) - other.tv_nsec as u32,
+                ))
+            }
+        } else {
+            None
+        }
+    }
+
+    pub fn checked_add_duration(&self, other: &Duration) -> Option<Timespec> {
+        let mut secs = self.tv_sec.checked_add(other.as_secs() as i64)?;
+
+        let mut nsec = other.subsec_nanos() + self.tv_nsec as u32;
+        if nsec >= NSEC_PER_SEC as u32 {
+            nsec -= NSEC_PER_SEC as u32;
+            secs = secs.checked_add(1)?;
+        }
+
+        Some(Timespec {
+            tv_sec: secs,
+            tv_nsec: nsec as _,
+        })
+    }
+
+    pub fn checked_sub_duration(&self, other: &Duration) -> Option<Timespec> {
+        let mut secs = self.tv_sec.checked_add(other.as_secs() as i64)?;
+
+        let mut nsec = self.tv_nsec as i32 - other.subsec_nanos() as i32;
+        if nsec < 0 {
+            nsec += NSEC_PER_SEC as i32;
+            secs = secs.checked_sub(1)?;
+        }
+
+        Some(Timespec {
+            tv_sec: secs,
+            tv_nsec: nsec as _,
+        })
+    }
+}
+
+lazy_static! {
+    static ref VDSO: Vdso = Vdso::new().unwrap();
+}
+
+fn now(clockid: vdso_time::clockid_t) -> Timespec {
+    let mut tp = vdso_time::timespec {
+        tv_sec: 0,
+        tv_nsec: 0,
+    };
+    VDSO.clock_gettime(clockid, &mut tp).unwrap();
```

Comment:
This is a problem left over by history, since the vdso-time crate is serving for libos's linux API syscalls.
I agree to refactor the API to `Duration`.